### PR TITLE
t/11: Introduced the Emitter mixin with the events API

### LINKED ISSUE #11: Introduce the Events API
According to the following talk: https://github.com/ckeditor/ckeditor5-design/issues/7

### PR DESCRIPTION
**Pull request by fredck**
Fixes #11.

We still may have discussion around this topic, so changes may (or may not still happen).

Tests included. As of today, to run them, the `ckeditor5` repo must be a t/1.

### CODE REVIEW

File: src/emitter.js [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+		 * @param {Number} [priority] The priority of this callback in relation to other callbacks to that same event.
+		 * Lower values are called first. Defaults to `10`.
+		 */
+		on: function( event, callback, ctx, priority ) {
+			var callbacks = getCallbacks( this, event );
+
+			var wasAdded;
+
+			// Priority defaults to 10.
+			if ( typeof priority != 'number' ) {
```
Comment (gregpabian) [inline src/emitter.js]: Right nowe we have `utils.isObject()` and `utils.isFunction()`, so why not introducing `isNumber()` and similar and use it here?

Comment (fredck) [inline src/emitter.js]: I was wondering if this is worth for such kind of basic checks.

Btw, we have isObject() and isFunction() not because we need them in general, just because utils itself needs it.

In fact, we're suppose to remove such functions from utils from now with #7. I'm just waiting for #4 to get merged to propose this change there (it's already pushed in t/7).

File: src/emitter.js [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+			this.on( event, onceCallback, ctx, priority );
+		},
+
+		/**
+		 * Stops executing the callback on the given event.
+		 *
+		 * @param {String} event The name of the event.
+		 * @param {Function} callback The function to stop being called.
+		 */
+		off: function( event, callback ) {
```
Comment (gregpabian) [inline src/emitter.js]: What if we had the same callback but executed in different contexts? This would remove all those handlers. I think we need an optional context to be passed.

File: src/emitter.js [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+			var eventCallbacks = emitterInfo && event && emitterInfo.callbacks[ event ];
+
+			// Stop if nothing has been listened.
+			if ( !emitters || ( emitter && !emitterInfo ) || ( event && !eventCallbacks ) ) {
+				return;
+			}
+
+			if ( callback ) {
+				// All params provided. off() that single callback.
+				emitter.off( event, callback );
```
Comment (gregpabian) [inline src/emitter.js]: Refers to my previous comment - this will remove all matching callbacks - no matter what's the context.

File: src/emitter.js [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+			} else if ( eventCallbacks ) {
+				// Only emitter and event provided. off() all callbacks for that event.
```
Comment (gregpabian) [inline src/emitter.js]: Missing articles.

File: src/emitter.js [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+				while ( ( callback = eventCallbacks.pop() ) ) {
+					emitter.off( event, callback );
+				}
+				delete emitterInfo.callbacks[ event ];
+			} else if ( emitterInfo ) {
+				// Only emitter provided. off() all events for that emitter.
```
Comment (gregpabian) [inline src/emitter.js]: Missing articles.

File: src/emitter.js [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+				// No params provided. off() all emitters.
+				for ( emitterId in emitters ) {
+					this.stopListening( emitters[ emitterId ].emitter );
+				}
+				delete this._listeningTo;
+			}
+		},
+
+		/**
+		 * Fires and event, executing all callbacks registered for it.
```
Comment (gregpabian) [inline src/emitter.js]: "Fires an~~d~~ event (...)"

File: src/emitter.js [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+		 *
+		 * The first parameter passed to callbacks is a {EventInfo} object, followed by the optional `args` provided in
```
Comment (gregpabian) [inline src/emitter.js]: "a**n** {EventInfo} (...)"

File: src/emitter.js [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+			var eventInfo = new EventInfo( event );
+
+			// Take the list of arguments to pass to the callbacks.
+			args = Array.prototype.slice.call( arguments, 1 );
+			args.unshift( eventInfo );
+
+			for ( var i = 0; i < callbacks.length; i++ ) {
+				callbacks[ i ].callback.apply( callbacks[ i ].ctx, args );
+
+				if ( eventInfo.stop.called ) {
```
Comment (gregpabian) [inline src/emitter.js]: Add a comment this one please.

File: src/emitter.js [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+					break;
+				}
+
+				if ( eventInfo.off.called ) {
```
Comment (gregpabian) [inline src/emitter.js]: Add a comment this one please.

File: tests/emitter/emitter.js [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+
+		emitter.fire( 'test' );
+		emitter.fire( 'test' );
+
+		sinon.assert.calledThrice( spy1 );
+		sinon.assert.calledOnce( spy2 );
+		sinon.assert.calledThrice( spy3 );
+	} );
+
+	it( 'should no fail with unknown events', function() {
```
Comment (gregpabian) [inline tests/emitter/emitter.js]: "should no**t** fail (...)"

File: tests/emitter/emitter.js [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+		emitter.off( 'test', function() {} );
+	} );
+} );
+
+describe( 'listenTo', function() {
```
Comment (gregpabian) [inline tests/emitter/emitter.js]: I miss a test for calling `listenTo` on non-emitter target.

File: src/emitter.js [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+					emitter.off( event, callback );
+				}
+				delete emitterInfo.callbacks[ event ];
+			} else if ( emitterInfo ) {
+				// Only emitter provided. off() all events for that emitter.
+				for ( event in emitterInfo.callbacks ) {
+					this.stopListening( emitter, event );
+				}
+				delete emitters[ emitterId ];
+			} else {
```
Comment (gregpabian) [inline src/emitter.js]: And what if we do the following:
1. Call listenTo with emitterA
2. Call stopListening with emitterB that we've never listened to

:trollface: ??

As a result we'll stop listening to the emitterA and every other emitter.

File: tests/emitter/emitter.js [hunk 13]
```diff
@@ -0,0 +1,356 @@
...
+		emitter.fire( 'test' );
+
+		sinon.assert.called( spy );
+	} );
+} );
+
+describe( 'stopListening', function() {
+	beforeEach( refreshListener );
+
+	it( 'should stop listening callback on event', function() {
```
Comment (gregpabian) [inline tests/emitter/emitter.js]: I'd rephrase this one:
"should stop listening to a specific event calback"

And besides I'm missing "listening **to**" in every other test case.

File: tests/emitter/emitter.js [hunk 14]
```diff
@@ -0,0 +1,356 @@
...
+
+		listener.listenTo( emitter, 'test', spy );
+
+		emitter.fire( 'test' );
+
+		sinon.assert.called( spy );
+	} );
+} );
+
+describe( 'stopListening', function() {
```
Comment (gregpabian) [inline tests/emitter/emitter.js]: Needs a test case for my trolly scenario: https://github.com/ckeditor/ckeditor5-core/pull/12/files#r24569343

File: tests/eventinfo/eventinfo.js [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+		var EventInfo = modules.eventinfo;
+
+		var event = new EventInfo( 'test' );
+
+		expect( event.name ).to.equals( 'test' );
+		expect( event.stop.called ).to.not.be.true();
+		expect( event.off.called ).to.not.be.true();
+	} );
+
+	it( 'should have cancel() and off() marked', function() {
```
Comment (gregpabian) [inline tests/eventinfo/eventinfo.js]: I don't see `cancel` anywhere. Can you update this description?

File: tests/utils/utils.js [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+
+var modules = bender.amd.require( 'utils' );
+
+describe( 'spy', function() {
+	it( 'should register calls', function() {
+		var utils = modules.utils;
+
+		var fn1 = utils.spy();
+		var fn2 = utils.spy();
+
```
Comment (gregpabian) [inline tests/utils/utils.js]: I think you miss a check if the `fn1` had the `called` property set to `false` after initialization.

File: src/emitter.js (hunk 1 above)
Comment (Reinmar) [inline src/emitter.js]: I don't see a reason why we should replace `typeof` with a method call doing exactly the same.  Tools functions make sense only if a check is complicated or isn't straightforward (like the `Object.prototype.toString.call` hack). Checking if something is a number is a basic operation and we don't care about `new Number( 5 )` or `NaN`. Replacing everything with method calls is a road to having `assign( left, right )` ;).

File: src/emitter.js [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+	var EmmitterMixin = {
+		/**
+		 * Registers a callback function to be executed when an event is fired.
+		 *
+		 * @param {String} event The name of the event.
+		 * @param {Function} callback The function to be called on event.
+		 * @param {Object} [ctx] The object that represents `this` in the callback. Defaults to the object firing the
+		 * event.
+		 * @param {Number} [priority] The priority of this callback in relation to other callbacks to that same event.
+		 * Lower values are called first. Defaults to `10`.
```
Comment (Reinmar) [inline src/emitter.js]: I never liked the default == 10. It's like you have a space 0-9 to do something earlier, and a lot more space after that. But then you end up with -1, -2, etc., because with time it turns out that you need more... I feel that it will be cleaner to have default == 0.

File: src/emitter.js (hunk 1 above)
Comment (gregpabian) [inline src/emitter.js]: So you want to do `obj !== null && typeof obj == 'object'` every time? Your call... I just want to stick to one convention, cause mix of typeofs and utils methods doesn't look so great (at least for me).

File: src/emitter.js (hunk 2 above)
Comment (Reinmar) [inline src/emitter.js]: Totally +1. Please also see https://github.com/ckeditor/ckeditor5-design/issues/7#issuecomment-74497143.

File: src/emitter.js (hunk 1 above)
Comment (Reinmar) [inline src/emitter.js]: The object case is a different case. I don't think that it makes sense to use `isNumber` and `isBoolean` everywhere just because `isObject` and `isRegExp` exist. It can really lead to `isNull` functions. But on the other hand I agree that having everything consistent is cool too, so actually part of me wants the `isNumber` method :D.

File: src/emitter.js [hunk 18]
```diff
@@ -0,0 +1,275 @@
...
+			}
+
+			callback = {
+				callback: callback,
+				ctx: ctx || this,
+				priority: priority
+			};
+
+			// Add the callback to the list in the right priority position.
+			for ( var i = 0; i < callbacks.length; i++ ) {
```
Comment (Reinmar) [inline src/emitter.js]: The reverse priority search from https://github.com/ckeditor/ckeditor-dev/blob/master/core/event.js#L174 actually makes sense, because most listeners have the default priority.

Comment (Reinmar) [inline src/emitter.js]: Fixed with 939f5d2.

File: src/emitter.js (hunk 3 above)
Comment (Reinmar) [inline src/emitter.js]: I'm for making the `off()` method more precise, but I wouldn't do that with `stopListening()`. Actually, I would even remove the `callback` (and maybe even the `event`) param and made it a function to "stop listening to: a) anything, b) given emitter)". If you need to be more precise just use `off()`. This should reduce `listenTo()` and `stopListening()` complexity which currently is pretty high and I'm not sure if we know that we need those features.

So the question is - are there any clear cases when `event` and `callback` may be useful and `off()` is not a solution?

File: src/emitter.js (hunk 5 above)
Comment (Reinmar) [inline src/emitter.js]: Fixed with 61a8cfb.

File: src/emitter.js (hunk 4 above)
Comment (Reinmar) [inline src/emitter.js]: Fixed with 61a8cfb.

File: src/emitter.js (hunk 17 above)
Comment (fredck) [inline src/emitter.js]: I think it's more a backwards alignment, to not have to change everything in the API (even principles) for the sake of reviewing every single thing.

Ofc, if there is a general consensus we can change it, because it's the last occasion for it.

File: src/emitter.js (hunk 6 above)
Comment (Reinmar) [inline src/emitter.js]: 41ad6dd

File: src/emitter.js (hunk 7 above)
Comment (Reinmar) [inline src/emitter.js]: 41ad6dd

File: src/emitter.js (hunk 8 above)
Comment (Reinmar) [inline src/emitter.js]: I tried really hard but it is pretty straightforward :P 41ad6dd.

File: src/emitter.js (hunk 9 above)
Comment (Reinmar) [inline src/emitter.js]: Thing that worries me is that if `stop()` was called, then `off()` call will be ignored. It's not what I would expect.

File: src/emitter.js (hunk 17 above)
Comment (Reinmar) [inline src/emitter.js]: I think that there's already enough API changes to force us to review every single line :D. But I don't also have a strong opinion about the default priority, so I'm curious if anyone else would like to see this change.

Comment (mlewand) [inline src/emitter.js]: As for me default priority 10 is fine. Listener priority is a kind of our thing so there is so we are free to use CKE4 value. It will be less confusing for developers to switch from 4 to 5.

File: src/emitter.js [hunk 19]
```diff
@@ -0,0 +1,279 @@
...
+		 * @param {Function} callback The function to be called on event.
+		 * @param {Object} [ctx] The object that represents `this` in the callback. Defaults to the object firing the
+		 * event.
+		 * @param {Number} [priority] The priority of this callback in relation to other callbacks to that same event.
+		 * Lower values are called first. Defaults to `10`.
+		 */
+		on: function( event, callback, ctx, priority ) {
+			var callbacks = getCallbacks( this, event );
+
+			// Priority defaults to 10.
```
Comment (mlewand) [inline src/emitter.js]: This is just super minor thing, but couldn't we just write: "Priority default value."? That way you reduce the DRY factor and if by any means default value would change, you reduce the risk that the comment will be outdated.

I do belive that we're very unlikely to change this one, but it's a general suggestion.

File: src/emitter.js (hunk 17 above)
Comment (gregpabian) [inline src/emitter.js]: BTW isn't there a syntax for the default value? Something like:
```
 @param {Number} [priority=10] The priority of this callback in relation to other callbacks to that same event.
```

That's how it looks in JSDoc at least.

Comment (Reinmar) [inline src/emitter.js]: > BTW isn't there a syntax for the default value? Something like:

This is also how it works in JSDuck.

File: tests/emitter/emitter.js (hunk 11 above)
Comment (fredck) [inline tests/emitter/emitter.js]: What's the point of this test? Ensure that it fails?

Comment (Reinmar) [inline tests/emitter/emitter.js]: It's like a nanny for kids - to ensure they don't do something stupid :D.

Comment (gregpabian) [inline tests/emitter/emitter.js]: :D

File: tests/emitter/emitter.js (hunk 14 above)
Comment (fredck) [inline tests/emitter/emitter.js]: Is this what you meant for this?

```js
it( 'should not stop other emitters when a non-listened emitter is provided', function() {
	var spy = sinon.spy();

	var emitter1 = getEmitterInstance();
	var emitter2 = getEmitterInstance();

	listener.listenTo( emitter1, 'test', spy );

	listener.stopListening( emitter2 );

	emitter1.fire( 'test' );

	sinon.assert.called( spy );
} );
```

It goes green for me, but from your comments I had the impression that you expected it to be red. Maybe you want a different test. Lemme know.

Comment (gregpabian) [inline tests/emitter/emitter.js]: I was thinking about a similar case, but now I see that this part actually covers it.
```js
if ( !emitters || ( emitter && !emitterInfo ) || ( event && !eventCallbacks ) ) {
	return;
}
```

So I think we don't need this case.

File: tests/emitter/emitter.js (hunk 11 above)
Comment (fredck) [inline tests/emitter/emitter.js]: Nah... we would have to create test for failure for every single feature so, which is pointless. :P

File: src/emitter.js (hunk 3 above)
Comment (fredck) [inline src/emitter.js]: Well, for my understanding, the only difference between `listenTo()` and `on()` is who's taking the responsibility of removing the callbacks later. One is the listener, the other is the emitter. Therefore I think that they fit for the very same scenarios.

Anyway, those methods are already in place, paired with tests. They're not that complex... it's more about a series of if's checking parameters. I would take them as good for now.

File: tests/emitter/emitter.js (hunk 14 above)
Comment (fredck) [inline tests/emitter/emitter.js]: Anyway, I've ended up pushing this tests anyway, as it was already in place. It's not a big deal.

**Comment by fredck** [comment]
All comments addressed. Ready for review again.

File: src/emitter.js (hunk 3 above)
Comment (Reinmar) [inline src/emitter.js]: > Therefore I think that they fit for the very same scenarios.

And what those scenarios are? I haven't worked with Backbone, but AFAIK 90% of cases are `obj.stopListening()`. If we are coding stuff which is not needed we are making a big mistake.

Besides that, I still don't like passing all the arguments to `off()` and even more to `stopListening()`, because not only this is long, but also problematic to do and maintain. If we make `on()` returning a listener instance (which is basically this [object](https://github.com/ckeditor/ckeditor5-core/pull/12/files#diff-d28ecdbfa507067c55c71c8d7c12167aR35) on steroids) we can IMO greatly improve the API.

> Anyway, those methods are already in place, paired with tests.

TBH, that should not be an argument. This way we will accept everything if it's already coded and working.

> They're not that complex... it's more about a series of if's checking parameters.

And these series are IMO unnecessarily complex. Plus, there's that big private object `this._listeningTo` which wouldn't be necessary if we resigned from the listeningTo's params and made `on()` return listener instance. We can make this cleaner and faster with little effort.

Comment (Reinmar) [inline src/emitter.js]: PS. To not stop this PR any more we can accept it and I'll propose my ideas in the next one. For the time being the stopListening() params should not be used.

Comment (fredck) [inline src/emitter.js]: > And what those scenarios are?  

Just like a button click. You can do button.on( ‘click’ ) or me.listenTo( button, ‘click’ ). Then I want to listen to other stuff from button, but during the execution lifecycle I may decide to stop listening to ‘click’ only… and at the end of the execution I could stop listening to everything.

> TBH, that should not be an argument. This way we will accept everything if it's already coded and working.

In this case we’re talking about a pretty standard API that we “expect” to be used. We’re not inventing anything new here. That’s the way people are used to have such kind of API out there, where you pass the same arguments to listen and stop listening to events.

> And these series are IMO unnecessarily complex.

Well, they’re the very minimum needed to have those feature available.

Ofc if the above arguments don’t convince we can remove features and code (and risk to have to get then back later).

Still the idea of having a `{ off() }` object returned by `on()` sounds good and it definitely fits a new issue.

Comment (Reinmar) [inline src/emitter.js]: > Just like a button click. You can do button.on( ‘click’ ) or me.listenTo( button, ‘click’ ). Then I want to listen to other stuff from button, but during the execution lifecycle I may decide to stop listening to ‘click’ only… and at the end of the execution I could stop listening to everything.

The thing is that I would love to know that I'm removing exactly the listener __I__ added. If you do `me.stopListening( button, 'click' )` you know that you stopped listening to "click", but what if a child or parent class started listening as well? Booom, happy debugging ;P. Ofc, you can pass the callback too or even the context and priority, but then it's awful code.

That's why I find `listener.off()` much nicer. You store the listener and remove just it.

> In this case we’re talking about a pretty standard API that we “expect” to be used.

But is it API that we want people to use? I don't want therefore I'll open a new issue for that.

File: src/emitter.js [hunk 20]
```diff
@@ -0,0 +1,284 @@
...
+		off: function( event, callback, ctx ) {
+			var callbacks = getCallbacksIfAny( this, event );
+
+			if ( !callbacks ) {
+				return;
+			}
+
+			for ( var i = 0; i < callbacks.length; i++ ) {
+				if ( callbacks[ i ].callback == callback ) {
+					if ( !ctx || ctx == callbacks[ i ].ctx ) {
```
Comment (gregpabian) [inline src/emitter.js]: Why separate `if`?

Comment (fredck) [inline src/emitter.js]: Just to have cleaner/readable code and avoid long if's.

We can change it if it looks so bad.
